File: src/jsMain/kotlin/com/scoreo/infrastructure/google/GoogleDriveClient.kt
package com.scoreo.infrastructure.google

import com.scoreo.domain.port.SyncException
import kotlinx.browser.window
import kotlinx.coroutines.await
import kotlinx.coroutines.delay

private const val DRIVE_API_BASE = "https://www.googleapis.com/drive/v3/files"

class GoogleDriveClient(
    private val getToken: suspend () -> String?,
) {
    suspend fun findFile(fileName: String): Result<String?> {
        val token = getToken() ?: return Result.failure(SyncException.NotAuthenticated)
        val escapedName = fileName.replace("\\", "\\\\").replace("'", "\\'")
        val url = "$DRIVE_API_BASE?spaces=appDataFolder&q=name='$escapedName'&fields=files(id,name)"
        return retryWithBackoff {
            try {
                val body = asyncGet(url, token)
                val json = JSON.parse(body) as FilesResponse
                Result.success(json.files?.firstOrNull()?.id)
            } catch (e: SyncException) {
                Result.failure(e)
            } catch (e: Exception) {
                Result.failure(SyncException.NetworkError)
            }
        }
    }

    suspend fun createFile(fileName: String, content: String, mimeType: String = "application/json"): Result<String> {
        val token = getToken() ?: return Result.failure(SyncException.NotAuthenticated)
        val metadata = """{ "name": "$fileName", "parents": ["appDataFolder"], "mimeType": "$mimeType" }"""
        val boundary = "scoreo_boundary"
        val body = buildMultipartBody(boundary, metadata, content, mimeType)
        val url = "$DRIVE_API_BASE?spaces=appDataFolder&uploadType=multipart"
        return retryWithBackoff {
            try {
                val responseBody = asyncPost(url, token, body, "multipart/related; boundary=$boundary")
                val json = JSON.parse(responseBody) as FileResponse
                Result.success(json.id ?: error("No id in response"))
            } catch (e: SyncException) {
                Result.failure(e)
            } catch (e: Exception) {
                Result.failure(SyncException.NetworkError)
            }
        }
    }

    suspend fun updateFile(fileId: String, content: String, mimeType: String = "application/json"): Result<String> {
        val token = getToken() ?: return Result.failure(SyncException.NotAuthenticated)
        val url = "$DRIVE_API_BASE/$fileId?spaces=appDataFolder&uploadType=media"
        return retryWithBackoff {
            try {
                asyncPatch(url, token, content, mimeType)
                Result.success(fileId)
            } catch (e: SyncException) {
                Result.failure(e)
            } catch (e: Exception) {
                Result.failure(SyncException.NetworkError)
            }
        }
    }

    suspend fun readFile(fileId: String): Result<String> {
        val token = getToken() ?: return Result.failure(SyncException.NotAuthenticated)
        val url = "$DRIVE_API_BASE/$fileId?alt=media"
        return retryWithBackoff {
            try {
                val body = asyncGet(url, token)
                Result.success(body)
            } catch (e: SyncException) {
                Result.failure(e)
            } catch (e: Exception) {
                Result.failure(SyncException.NetworkError)
            }
        }
    }

    suspend fun upsertFile(fileName: String, content: String, mimeType: String = "application/json"): Result<String> {
        val findResult = findFile(fileName)
        return findResult.fold(
            onSuccess = { fileId ->
                if (fileId != null) updateFile(fileId, content, mimeType)
                else createFile(fileName, content, mimeType)
            },
            onFailure = { throwable -> Result.failure(throwable) },
        )
    }

    // ── Retry ──

    private suspend fun <T> retryWithBackoff(
        maxRetries: Int = 3,
        initialDelayMs: Long = 1_000L,
        block: suspend () -> Result<T>,
    ): Result<T> {
        var delayMs = initialDelayMs
        var attempt = 0
        while (true) {
            val result = block()
            val failure = result.exceptionOrNull()
            if (failure == null || attempt >= maxRetries) return result
            when (failure) {
                is SyncException.RateLimited,
                is SyncException.NetworkError -> {
                    delay(delayMs)
                    delayMs *= 2
                    attempt++
                }
                else -> return result
            }
        }
    }

    // ── Private helpers ──

    private suspend fun asyncGet(url: String, token: String): String {
        val headers: dynamic = js("({})")
        headers["Authorization"] = "Bearer $token"
        val options: dynamic = js("({})")
        options.method = "GET"
        options.headers = headers
        val response = window.fetch(url, options).await()
        val body = response.text().await()
        return checkResponse(response.status.toInt(), body)
    }

    private suspend fun asyncPost(url: String, token: String, body: String, contentType: String): String {
        val headers: dynamic = js("({})")
        headers["Authorization"] = "Bearer $token"
        headers["Content-Type"] = contentType
        val options: dynamic = js("({})")
        options.method = "POST"
        options.headers = headers
        options.body = body
        val response = window.fetch(url, options).await()
        val responseBody = response.text().await()
        return checkResponse(response.status.toInt(), responseBody)
    }

    private suspend fun asyncPatch(url: String, token: String, body: String, contentType: String): String {
        val headers: dynamic = js("({})")
        headers["Authorization"] = "Bearer $token"
        headers["Content-Type"] = contentType
        val options: dynamic = js("({})")
        options.method = "PATCH"
        options.headers = headers
        options.body = body
        val response = window.fetch(url, options).await()
        val responseBody = response.text().await()
        return checkResponse(response.status.toInt(), responseBody)
    }

    private fun checkResponse(status: Int, body: String): String {
        return when {
            status in 200..299 -> body
            status == 401 -> throw SyncException.NotAuthenticated
            status == 429 -> throw SyncException.RateLimited
            status in 500..599 -> throw SyncException.NetworkError
            else -> throw SyncException.ApiError(status, body)
        }
    }

    private fun buildMultipartBody(boundary: String, metadata: String, content: String, mimeType: String): String {
        return """--$boundary
Content-Type: application/json; charset=UTF-8

$metadata
--$boundary
Content-Type: $mimeType

$content
--$boundary--"""
    }
}

// ── External JS types ──

private external interface FilesResponse {
    val files: Array<FileInfo>?
}

private external interface FileInfo {
    val id: String
    val name: String
}

private external interface FileResponse {
    val id: String?
}

private external object JSON {
    fun parse(text: String): dynamic
}

@Suppress("UNUSED")
private inline fun <reified T> JSON.parse(text: String): T =
    (this.parse(text) as T)

File: apps/scoreo/src/infrastructure/google/googleDriveClient.ts
import type { SyncException } from '../../domain/port/cloudSyncRepository'
import { err, ok, type Result } from '../../domain/result'

const DRIVE_API_BASE = 'https://www.googleapis.com/drive/v3/files'
/** Drive v3 requires this dedicated /upload/ prefix for any request that carries file content (multipart or media uploadType). */
const DRIVE_UPLOAD_API_BASE = 'https://www.googleapis.com/upload/drive/v3/files'

interface FileInfo {
  id: string
  name: string
}

interface FilesResponse {
  files?: FileInfo[]
}

interface FileResponse {
  id?: string
}

function isSyncException(value: unknown): value is SyncException {
  return typeof value === 'object' && value !== null && 'kind' in value
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

/** Structural shape of GoogleDriveClient's public API, so tests can substitute a fake without extending the real class. */
export interface DriveClient {
  findFile(fileName: string): Promise<Result<string | undefined, SyncException>>
  createFile(fileName: string, content: string, mimeType?: string): Promise<Result<string, SyncException>>
  updateFile(fileId: string, content: string, mimeType?: string): Promise<Result<string, SyncException>>
  readFile(fileId: string): Promise<Result<string, SyncException>>
  upsertFile(fileName: string, content: string, mimeType?: string): Promise<Result<string, SyncException>>
}

export class GoogleDriveClient implements DriveClient {
  constructor(private readonly getToken: () => Promise<string | undefined>) {}

  async findFile(fileName: string): Promise<Result<string | undefined, SyncException>> {
    const token = await this.getToken()
    if (!token) return err({ kind: 'NotAuthenticated', message: 'Not authenticated' })
    const escapedName = fileName.replace(/\\/g, '\\\\').replace(/'/g, "\\'")
    const url = `${DRIVE_API_BASE}?spaces=appDataFolder&q=name='${escapedName}'&fields=files(id,name)`
    return this.retryWithBackoff(async () => {
      try {
        const body = await this.asyncGet(url, token)
        const json = JSON.parse(body) as FilesResponse
        return ok(json.files?.[0]?.id)
      } catch (e) {
        return err(isSyncException(e) ? e : { kind: 'NetworkError' })
      }
    })
  }

  async createFile(
    fileName: string,
    content: string,
    mimeType = 'application/json',
  ): Promise<Result<string, SyncException>> {
    const token = await this.getToken()
    if (!token) return err({ kind: 'NotAuthenticated', message: 'Not authenticated' })
    const metadata = `{ "name": "${fileName}", "parents": ["appDataFolder"], "mimeType": "${mimeType}" }`
    const boundary = 'scoreo_boundary'
    const body = this.buildMultipartBody(boundary, metadata, content, mimeType)
    const url = `${DRIVE_UPLOAD_API_BASE}?spaces=appDataFolder&uploadType=multipart`
    return this.retryWithBackoff(async () => {
      try {
        const responseBody = await this.asyncPost(url, token, body, `multipart/related; boundary=${boundary}`)
        const json = JSON.parse(responseBody) as FileResponse
        if (!json.id) throw new Error('No id in response')
        return ok(json.id)
      } catch (e) {
        return err(isSyncException(e) ? e : { kind: 'NetworkError' })
      }
    })
  }

  async updateFile(
    fileId: string,
    content: string,
    mimeType = 'application/json',
  ): Promise<Result<string, SyncException>> {
    const token = await this.getToken()
    if (!token) return err({ kind: 'NotAuthenticated', message: 'Not authenticated' })
    const url = `${DRIVE_UPLOAD_API_BASE}/${fileId}?spaces=appDataFolder&uploadType=media`
    return this.retryWithBackoff(async () => {
      try {
        await this.asyncPatch(url, token, content, mimeType)
        return ok(fileId)
      } catch (e) {
        return err(isSyncException(e) ? e : { kind: 'NetworkError' })
      }
    })
  }

  async readFile(fileId: string): Promise<Result<string, SyncException>> {
    const token = await this.getToken()
    if (!token) return err({ kind: 'NotAuthenticated', message: 'Not authenticated' })
    const url = `${DRIVE_API_BASE}/${fileId}?alt=media`
    return this.retryWithBackoff(async () => {
      try {
        const body = await this.asyncGet(url, token)
        return ok(body)
      } catch (e) {
        return err(isSyncException(e) ? e : { kind: 'NetworkError' })
      }
    })
  }

  async upsertFile(
    fileName: string,
    content: string,
    mimeType = 'application/json',
  ): Promise<Result<string, SyncException>> {
    const findResult = await this.findFile(fileName)
    if (!findResult.ok) return err(findResult.error)
    return findResult.value ? this.updateFile(findResult.value, content, mimeType) : this.createFile(fileName, content, mimeType)
  }

  private async retryWithBackoff<T>(
    block: () => Promise<Result<T, SyncException>>,
    maxRetries = 3,
    initialDelayMs = 1_000,
  ): Promise<Result<T, SyncException>> {
    let delayMs = initialDelayMs
    let attempt = 0
    for (;;) {
      const result = await block()
      if (result.ok || attempt >= maxRetries) return result
      if (result.error.kind === 'RateLimited' || result.error.kind === 'NetworkError') {
        await sleep(delayMs)
        delayMs *= 2
        attempt++
      } else {
        return result
      }
    }
  }

  private async asyncGet(url: string, token: string): Promise<string> {
    const response = await fetch(url, {
      method: 'GET',
      headers: { Authorization: `Bearer ${token}` },
    })
    const body = await response.text()
    return this.checkResponse(response.status, body)
  }

  private async asyncPost(url: string, token: string, body: string, contentType: string): Promise<string> {
    const response = await fetch(url, {
      method: 'POST',
      headers: { Authorization: `Bearer ${token}`, 'Content-Type': contentType },
      body,
    })
    const responseBody = await response.text()
    return this.checkResponse(response.status, responseBody)
  }

  private async asyncPatch(url: string, token: string, body: string, contentType: string): Promise<string> {
    const response = await fetch(url, {
      method: 'PATCH',
      headers: { Authorization: `Bearer ${token}`, 'Content-Type': contentType },
      body,
    })
    const responseBody = await response.text()
    return this.checkResponse(response.status, responseBody)
  }

  private checkResponse(status: number, body: string): string {
    if (status >= 200 && status <= 299) return body
    if (status === 401) throw { kind: 'NotAuthenticated', message: 'Not authenticated' } satisfies SyncException
    if (status === 429) throw { kind: 'RateLimited' } satisfies SyncException
    if (status >= 500 && status <= 599) throw { kind: 'NetworkError' } satisfies SyncException
    throw { kind: 'ApiError', code: status, message: body } satisfies SyncException
  }

  private buildMultipartBody(boundary: string, metadata: string, content: string, mimeType: string): string {
    return `--${boundary}
Content-Type: application/json; charset=UTF-8

${metadata}
--${boundary}
Content-Type: ${mimeType}

${content}
--${boundary}--`
  }
}
